pkg/client/actionclient.go

```diff
@@ -180,6 +181,14 @@ func (c *actionClient) Upgrade(name, namespace string, chrt *chart.Chart, vals m
 	return rel, nil
 }
 
+func (c *actionClient) MarkFailed(rel *release.Release, reason string) error {
+	infoCopy := *rel.Info
+	releaseCopy := *rel
+	releaseCopy.Info = &infoCopy
+	releaseCopy.SetStatus(release.StatusFailed, reason)
+	return c.conf.Releases.Update(&releaseCopy)
+}
```

Comment:
While I see how it's convenient to put this functionality into the actionClient, I'm not convinced it makes a ton of sense otherwise. There's really only one way to mark a release as failed (this is it), so I'm wondering if we pull this back out of the action client interface and just put this logic directly into the reconciler.

The only missing piece I see for that is giving the Reconciler an `ActionConfigGetter` field, which would likely just involve adding the field, adding a `WithActionConfigGetter` functional option, and then tweaking the `addDefaults` function slightly to handle the fact that the reconciler may already have an `ActionConfigGetter` setup via the new functional option.

Reply:
Agree, extracted it and now head to adding a fake implementation for the `ActionConfigGetter`. After that it should be finished. The implementation is still a bit rough though.